### PR TITLE
use node version manager

### PR DESCRIPTION
**Pull request by krudos**

### CODE REVIEW

**Review by gengjiawen — CHANGES_REQUESTED** [review]
If you want to add a node version manager, I would prefer n. This would be more less intrusive and easy (use npm to install n)

File: Dockerfile [hunk 1]
```diff
@@ -87,12 +87,12 @@ RUN apt update -qq && apt install -qq -y --no-install-recommends \
...
 
-# install nodejs and yarn packages from nodesource
-RUN curl -sL https://deb.nodesource.com/setup_${NODE_VERSION} | bash - \
-    && apt-get update -qq \
-    && apt-get install -qq -y --no-install-recommends nodejs \
-    && npm i -g yarn \
-    && rm -rf /var/lib/apt/lists/*
+# install nodejs using n
+RUN curl -L https://raw.githubusercontent.com/tj/n/master/bin/n -o n \
+    && bash n lts \
```
Comment (gengjiawen) [inline Dockerfile:92]: why not just install Node.js once ? This would save us build time.

Comment (krudos) [reply Dockerfile]: done

File: Dockerfile [hunk 2]
```diff
@@ -87,12 +87,10 @@ RUN apt update -qq && apt install -qq -y --no-install-recommends \
...
-# install nodejs and yarn packages from nodesource
-RUN curl -sL https://deb.nodesource.com/setup_${NODE_VERSION} | bash - \
-    && apt-get update -qq \
-    && apt-get install -qq -y --no-install-recommends nodejs \
-    && npm i -g yarn \
-    && rm -rf /var/lib/apt/lists/*
+# install nodejs using n
+RUN curl -L https://raw.githubusercontent.com/tj/n/master/bin/n -o n \
+    && bash n $NODE_VERSION \
+    && npm install -g yarn
```
Comment (gengjiawen) [inline Dockerfile:93]: Nit: maybe add n here too since it will be in system path.

**Review by gengjiawen — APPROVED** [review]
LGTM

**Comment by cortinico** [comment]
I would refrain from adding a node version manager in the container frankly. Version managers are user tools. The idea behind a container is to have a fixed set of dependencies that are used to build a given tool.

**Comment by gengjiawen** [comment]
> I would refrain from adding a node version manager in the container frankly. Version managers are user tools. The idea behind a container is to have a fixed set of dependencies that are used to build a given tool.

It only added one npm package in the end, the cost is very small. In the long run, I think this is very useful for user-land to do matrix testing(Node.js latest, lts, minimum)

**Comment by leotm** [comment]
agree w a minimal container but also the small cost of `n` (or some other hip manager, not talking big o notation)

node 19's out (working for me) and others testing matrices ahead's good for long-term contribution (node 20 next easter)

noticed `ubuntu-22.04` comes ootb w: node 18, `n` and `nvm` (node 12 removed from all images last wk)

**Comment by gengjiawen** [comment]
Thx for contribution

**Pull request merged**